contracts/test/TreasuryVester.t.sol

```diff
@@ -0,0 +1,272 @@
+            DEFAULT_VESTING_END_TIMESTAMP
+        );
+    }
+
+    function testCreateVester() public {
+        lon.approve(address(treasuryVesterFactory), DEFAULT_VESTING_AMOUNT);
+        vm.expectEmit(
+            false, // We do not check the deployed Vester address as it's deterministic generated and irrelevant to correctness
+            true,
+            false,
+            true
+        );
+        emit VesterCreated(
+            address(this), // We do not check the deployed Vester address as it's deterministic generated and irrelevant to correctness
+            recipient,
+            DEFAULT_VESTING_AMOUNT
+        );
+        treasuryVester = TreasuryVester(
+            treasuryVesterFactory.createVester(
+                recipient,
+                DEFAULT_VESTING_AMOUNT,
+                DEFAULT_VESTING_BEGIN_TIMESTAMP,
+                DEFAULT_VESTING_CLIFF_TIMESTAMP,
+                DEFAULT_VESTING_END_TIMESTAMP
+            )
+        );
+        assertEq(treasuryVester.lon(), address(lon));
+        assertEq(treasuryVester.recipient(), recipient);
+        assertEq(treasuryVester.vestingAmount(), DEFAULT_VESTING_AMOUNT);
+        assertEq(treasuryVester.vestingBegin(), DEFAULT_VESTING_BEGIN_TIMESTAMP);
+        assertEq(treasuryVester.vestingCliff(), DEFAULT_VESTING_CLIFF_TIMESTAMP);
+        assertEq(treasuryVester.lastUpdate(), DEFAULT_VESTING_BEGIN_TIMESTAMP);
+    }
+
+    /*********************************************************
+     *              Test: set Vester recipient               *
+     *********************************************************/
+
+    function testCannotSetRecipientByNotRecipient() public {
+        lon.approve(address(treasuryVesterFactory), DEFAULT_VESTING_AMOUNT);
+        treasuryVester = TreasuryVester(
+            treasuryVesterFactory.createVester(
+                recipient,
+                DEFAULT_VESTING_AMOUNT,
+                DEFAULT_VESTING_BEGIN_TIMESTAMP,
+                DEFAULT_VESTING_CLIFF_TIMESTAMP,
+                DEFAULT_VESTING_END_TIMESTAMP
+            )
+        );
+        vm.expectRevert("unauthorized");
+        vm.prank(other);
+        treasuryVester.setRecipient(other);
+    }
+
+    function _getVestedAmount(
+        uint256 vestingAmount,
+        uint256 lastUpdate,
+        uint256 beginTimestamp,
+        uint256 endTimestamp
+    ) internal returns (uint256) {
+        return vestingAmount.mul(block.timestamp - lastUpdate).div(endTimestamp.sub(beginTimestamp));
+    }
+
+    function testVested() public {
+        lon.approve(address(treasuryVesterFactory), DEFAULT_VESTING_AMOUNT);
+        treasuryVester = TreasuryVester(
+            treasuryVesterFactory.createVester(
+                recipient,
+                DEFAULT_VESTING_AMOUNT,
+                DEFAULT_VESTING_BEGIN_TIMESTAMP,
+                DEFAULT_VESTING_CLIFF_TIMESTAMP,
+                DEFAULT_VESTING_END_TIMESTAMP
+            )
+        );
```

Comment:
nit: Maybe it's worth a util function to setup a valid vester

Reply:
Fixed in 66a1b2f

Reply:
@NIC619 I think approve can also include in the util. What do you think about it?

Reply:
I feel like it's not worth the readability improvement as much as creating a default vester? Also the function name would probably need to be changed to `_createVesterWithDefaultValuesAndApproveTreasuryFactory` as we want to be as explicit as possible in our tests and I don't not a fan of the name.😂 (maybe there's better name🤔)

Reply:
Actually what my original thought is not that complicate 😆 I was thinking something like

```solidity
function setupTreasuryVester(recipient, amount, begin, cliff, end) {
    lon.approve(address(treasuryVesterFactory), amount);
    return TreasuryVester(
            treasuryVesterFactory.createVester(
                recipient,
                amount,
                begin,
                cliff,
                end
            )
        );
}
```

But this is just a minor suggestion that we can take it not so serious. The original version it's ok for me too. We can just choose a preferable way for us to accomplish it 👍

Reply:
And I come up another idea that if we separate contract test cases into different files (as we discussed before), each file can have their custom before each hook. 

For these cases based on valid TreasuryVester, we can simply move more setup steps into setup function.

Just idea sharing, not to do right now 😆

Reply:
Yeah breaking tests into different files is definitely an good option if the test file grows too large.

Reply:
Let's settle with current version?

Reply:
OK, no problem 👍